File: September2022/C++/AverageOfLevelsInBinaryTree.cpp
// https://leetcode.com/problems/average-of-levels-in-binary-tree/
class Solution {
public:
    // TC: O(n)
    // SC: O(logn)
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> ans;
        queue<TreeNode*> queue;
        queue.push(root);
        
        while (!queue.empty()) {
            int size = queue.size();
            double sum = 0;
            for (int i = 0; i < size; i++) {
                TreeNode* node = queue.front();
                queue.pop();
                if (node->left != NULL) {
                    queue.push(node->left);
                }
                if (node->right != NULL) {
                    queue.push(node->right);
                }
                sum += node->val;
            }
            ans.push_back(sum / size);
        }
        
        return ans;
    }
};

File: September2022/Java/AverageOfLevelsInBinaryTree.java

// @saorav21994
// TC : O(n)
// SC : O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        
        List<Double> res = new ArrayList<>();
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        
        while (!queue.isEmpty()) {
            
            int size = queue.size();
            int csize = size;
            long sum = 0l;
            
            while (csize-- > 0) {
                TreeNode cur = queue.poll();
                sum += (long)cur.val;
                if (cur.left != null)
                     queue.offer(cur.left);
                if (cur.right != null) 
                    queue.offer(cur.right);
            }
            res.add((double)sum/(double)size);
            
        }
        
        return res;
        
        
    }
}
